README
MODULES
  Object::Base
    Multi-threaded base class to establish a class deriving relationship
    with parent classes

    See more <https://metacpan.org/pod/Object::Base>

  Object::Exception
    Multi-threaded base exception class

    See more <https://metacpan.org/pod/Object::Exception>

INSTALLATION
    To install this module type the following

            perl Makefile.PL
            make
            make test
            make install

    from CPAN

            cpan -i Object::Base

DEPENDENCIES
    This module requires these other modules and libraries:

    *   threads

    *   threads::shared

    *   forks

    *   SUPER

REPOSITORY
    GitHub <https://github.com/orkunkaraduman/p5-Object-Base>

    CPAN <https://metacpan.org/release/Object-Base>

AUTHOR
    Orkun Karaduman <orkunkaraduman@gmail.com>

COPYRIGHT AND LICENSE
    Copyright (C) 2017 Orkun Karaduman <orkunkaraduman@gmail.com>

    This program is free software: you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the License, or (at your
    option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
    Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program. If not, see <http://www.gnu.org/licenses/>.



Changes
1.14	2017-03-06
		- POD fixes
		- Minimum Perl version setted v5.8.8

1.13	2017-02-28
		- POD & README changes

1.12	2017-02-25
		- POD changes

1.11	2017-02-20
		- Fixed bug for threads/forks switch
		- Expanded tests

1.10	2017-02-16
		- Fix to 'Fails with all perls except 5.18.0..5.24.x'

1.09	2017-02-15
		- Object::Exception::dump_trace() indentation fix
		- Object::Exception msg attribute calling as method
		- Added $tracelevel in Object::Exception::throw()
		- $msg arg of Object::Exception::throw() can be as a object of child class Object::Exception

1.08	2017-02-12
		- Added traceback, dump_trace functions in Object::Exception
		- Attribute modifiers don't work with ':shared' feature

1.07	2017-02-11
		- Removed current_value in getter/setter

1.06	2017-02-10
		- Removed locks and key name check in Object::Base::TieHash::def() for performance
		- Added individual attribute locks
		- Improved general locks for thread-safety
		- Die with 'Object::Base can not be imported at run-time'
		- Calling Object::Base::attributes() at caller's attributes subroutine
		- Removed some quotemeta for perfomance and they are not needed
		- Added 'use SUPER;' for child classes
		- Added shared_clone for default values if it needs
		- Added 'not is_shared' condition at sub attr, FETCH, def
		- Added Object::Exception

1.05	2017-02-08
		- Added DESTROY method for calling SUPER in parent overriding

1.04	2017-02-07
		- Object-based attribute locks
		- Fix to modifiers

1.03	2017-02-06
		- Allowed attribute redefinition
		- Added default/getter/setter attribute definitions
		- Added ':lazy' special attribute
		- Added internal attribute locks for thread-safety
		- Added forks support for Perl not built to support threads

1.02	2017-02-01
		- Attribute name quotemeta fix
		- Added thread libraries in prerequirements

1.01	2017-01-31
		- ':shared' attribute shared_clone fix

1.00	2017-01-31
		- Object::Base first release


dev/test.pl
#! /usr/bin/perl
=head1 NAME

test.pl - for internal tests

=head1 VERSION

version not defined

=head1 ABSTRACT

for internal tests

=cut
use strict;
use warnings;
use v5.14;
use utf8;
use open qw(:utf8 :std);
use open IO => ':bytes';
use FindBin;
use Data::Dumper;

use lib "${FindBin::Bin}/../lib";


################################################################################
# Object::Base
################################################################################
	package Foo;
	use Object::Base;
	attributes ':shared', 'attr1', 'attr2';
	#
	package Bar;
	use Object::Base 'Foo';
	my $attr3_def = 6;
	my $attr3_val;
	attributes ':shared' => undef, 'attr2' => undef, ':lazy',
		'attr3' => {
			'default' => sub {
				my ($self, $key) = @_;
				print "> default key=$key\n";
				$attr3_val = $attr3_def-1;
				return $attr3_def;
			},
			'getter' => sub {
				my ($self, $key) = @_;
				print "> getter key=$key\n";
				return $attr3_val+1;
			},
			'setter' => sub {
				my ($self, $key, $value) = @_;
				print "> setter key=$key value=$value\n";
				$attr3_val = $value-1;
			},
		}
	;
	#
	package main;
	use threads;
	use threads::shared;
	#
	# object of Foo
	my $foo = Foo->new();
	#
	# special attribute ':shared'
	print "\$foo is ", is_shared($foo)? "shared": "not shared", "\n";
	#
	# usage of attribute
	$foo->attr1(1);
	print $foo->attr1, "\n"; # prints '1'
	#
	# attributes are lvalued
	$foo->attr1++;
	print $foo->attr1, "\n"; # prints '2'
	#
	# assigning ref values to shared class attributes
	eval { $foo->attr2 = { key1 => 'val1' } }; print "Eval: $@"; # prints error 'Eval: Invalid value for shared scalar at ...'
	$foo->attr2({ key2 => 'val2' }); # uses shared_clone assigning ref value
	print $foo->attr2->{key2}, "\n"; # prints 'val2'
	#
	# object of derived class Bar
	my $bar = Bar->new();
	#
	# features are overridable
	print "\$bar is ", is_shared($bar)? "shared": "not shared", "\n"; # prints '$bar is not shared'
	#
	# attributes can be added derived classes
	# attributes can have modifiers: default
	print "attr3 default value is ", $bar->attr3, "\n"; # prints 'attr3 default value is 6'
	#
	# attributes can have modifiers: setter
	$bar->attr3 = 3;
	print "attr3 set", "\n";
	#
	# attributes can have modifiers: getter
	print "attr3 value ", $bar->attr3, " and stored as $attr3_val", "\n"; # prints 'attr3 value 3 and stored as 2'
	#
	# attributes are inheritable
	$bar->attr1(3);
	#
	# attributes are overridable
	eval { $bar->attr2 = 4 }; print "Eval: $@"; # prints error 'Eval: Attribute attr2 is not defined in Bar at ...'
	#
	# attributes in thread
	my $thr1 = threads->create(sub { $foo->attr1 = 5; $bar->attr1 = 5; });
	my $thr2 = threads->create(sub { sleep 1; print "\$foo is shared and attr1: ", $foo->attr1, ", \$bar is not shared and attr1: ", $bar->attr1, "\n"; });
	# prints '$foo is shared and attr1: 5, $bar is not shared and attr1: 3'
	$thr1->join();
	$thr2->join();
################################################################################


################################################################################
# Object::Exception
################################################################################
	package SampleException;
	use Object::Base qw(Object::Exception);
	#
	package main;
	use Object::Exception;
	#
	# Enable DEBUG for traceback
	our $DEBUG = 1;
	#
	# throws Object::Exception type and its msg: Exception1
	eval
	{
		throw("Exception1");
	};
	if ($@)
	{
		warn $@ if ref($@) eq "Object::Exception";
	}
	#
	# throws SampleException type and its msg: This is sample exception
	sub sub_exception
	{
		SampleException->throw("This is sample exception");
	}
	eval
	{
		sub_exception();
	};
	if ($@) {
		# $@ and $@->message returns same result
		warn $@->message if ref($@) eq "SampleException";
	}
	#
	# throws Object::Exception type and its message: SampleException. Because msg is not defined!
	eval
	{
		SampleException->throw();
	};
	if ($@)
	{
		if (ref($@) eq "SampleException")
		{
			warn $@;
		} else
		{
			# warns 'This is type of Object::Exception and its message: SampleException'
			warn "This is type of ".ref($@)." and its message: $@";
		}
	}
################################################################################


say "OK";
exit 0;
__END__
=head1 AUTHOR

Orkun Karaduman <orkunkaraduman@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2017  Orkun Karaduman <orkunkaraduman@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;


WriteMakefile(
	NAME				=> 'Object::Base',
	VERSION				=> '1.14',
	MIN_PERL_VERSION	=> '5.008008',
	PREREQ_PM			=> {
		'threads'		=> '1.72',
		'threads::shared' => '1.28',
		'forks'			=> '0.29',
		'SUPER'			=> '1.16',
	},
	EXE_FILES			=> [qw(
	)],
	AUTHOR				=> 'Orkun Karaduman <orkunkaraduman@gmail.com>',
	ABSTRACT			=> 'Multi-threaded base class and exception class',
	LICENSE				=> 'gpl_3',
	META_MERGE			=> {
		'meta-spec'		=> {
			version		=> 2,
		},
		resources		=> {
			homepage	=> 'https://metacpan.org/release/Object-Base',
			license		=> 'https://www.gnu.org/licenses/gpl-3.0.html',
			repository	=> {
				type	=> 'git',
				url		=> 'git://git@github.com:orkunkaraduman/p5-Object-Base.git',
				web		=> 'https://github.com/orkunkaraduman/p5-Object-Base',
			},
			bugtracker	=> {
				web		=> 'https://github.com/orkunkaraduman/p5-Object-Base/issues',
				mailto	=> 'bug-Object-Base@rt.cpan.org',
			},
		},
	},
);


dev/dist.pl
#! /usr/bin/perl
=head1 NAME

dist.pl - distribution generator

=head1 VERSION

version not defined

=head1 ABSTRACT

distribution generator

=cut
use strict;
use warnings;
use v5.14;
use utf8;
use open qw(:utf8 :std);
use open IO => ':bytes';
use FindBin;
use Cwd;


#my $podPath = "lib/" . "Object::Base" =~ s/\:\:/\//gr . ".pm";
my $podPath = "README.pod";
my $base = "${FindBin::Bin}/..";
cwd($base);

system("perl Makefile.PL");
system("pod2markdown --html-encode-chars 1 $podPath > README.md");
system("pod2text $podPath > README");
system("rm MANIFEST; make manifest");
system("make dist");

exit 0;
__END__
=head1 AUTHOR

Orkun Karaduman <orkunkaraduman@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2017  Orkun Karaduman <orkunkaraduman@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut


lib/Object/Exception.pm
package Object::Exception;
=head1 NAME

Object::Exception - Multi-threaded exception class

=head1 VERSION

version 1.13

=head1 ABSTRACT

Multi-threaded exception class

	package SampleException;
	use Object::Base qw(Object::Exception);
	#
	package main;
	use Object::Exception;
	#
	# Enable DEBUG for traceback
	our $DEBUG = 1;
	#
	# throws Object::Exception type and its msg: Exception1
	eval
	{
		throw("Exception1");
	};
	if ($@)
	{
		warn $@ if ref($@) eq "Object::Exception";
	}
	#
	# throws SampleException type and its msg: This is sample exception
	sub sub_exception
	{
		SampleException->throw("This is sample exception");
	}
	eval
	{
		sub_exception();
	};
	if ($@) {
		# $@ and $@->message returns same result
		warn $@->message if ref($@) eq "SampleException";
	}
	#
	# throws Object::Exception type and its message: SampleException. Because msg is not defined!
	eval
	{
		SampleException->throw();
	};
	if ($@)
	{
		if (ref($@) eq "SampleException")
		{
			warn $@;
		} else
		{
			# warns 'This is type of Object::Exception and its message: SampleException'
			warn "This is type of ".ref($@)." and its message: $@";
		}
	}

=head1 DESCRIPTION

=head2 Functions

=head3 traceback($level)

returns array specified level of traceback by calling point of traceback function.

=head3 dump_trace(@trace)

returns string dump of trace array. Always ends with EOL ($/) generally "\n".

=head3 throw($msg)

dies with new Object::Exception instance with specified message.

=head2 Methods

=head3 $class->new($msg)

returns new Object::Exception instance with specified message. If $main::DEBUG is setted TRUE, $object->debug attribute is setted 1.

=head3 $object->message()

returns message of Object::Exception instance. If $msg is defined with new() or throw(), always ends with EOL ($/) generally "\n".
If $object->debug attribute is TRUE, dump generated with dump_trace is added to end of message.

=head3 $class->throw($msg)

dies with new Object::Exception derived-class instance with specified message. If B<$class> is not derived from Object::Exception,
does nothing. $msg value must be specified explicitly and it can be B<undef>. Otherwise, method runs as B<throw($class)> function.

=cut
use Object::Base qw(Exporter);
use overload '""' => \&message;


BEGIN
{
	require 5.008;
	$Object::Exception::VERSION = '1.13';
	@Object::Exception::EXPORT = qw(throw);
	@Object::Exception::EXPORT_OK = qw(traceback dump_trace);
}


attributes qw(:shared msg debug trace);


sub traceback
{
	my ($level) = @_;
	$level = 0 unless defined($level) and $level >= 0;
	my @result;
	while (scalar(my @caller = caller($level++)))
	{
		my @caller_next = caller($level);
		push @result, {
			package => $caller[0],
			filename => $caller[1],
			line => $caller[2],
			subroutine => $caller_next[3],
		};
	}
	return @result;
}

sub dump_trace
{
	local $/ = "\n" unless defined($/);
	my @trace = @_;
	my $result = "";
	my $i = 1;
	for my $trace (reverse @trace)
	{
		$result .= sprintf("%-".($i*1)."s", "");
		$result .= "in $trace->{package} ";
		$result .= "at ";
		$result .= "$trace->{subroutine} " if defined($trace->{subroutine});
		$result .= "$trace->{filename} ";
		$result .= "line $trace->{line}$/";
	} continue
	{
		$i++;
	}
	return $result;
}

sub throw
{
	my $class;
	if (@_ > 1)
	{
		$class = shift;
	} else
	{
		$class = __PACKAGE__;
	}
	my ($msg, $tracelevel) = @_;
	return unless defined($class) and not ref($class) and UNIVERSAL::isa($class, __PACKAGE__);
	my @trace;
	if (ref($msg))
	{
		return unless UNIVERSAL::isa($msg, __PACKAGE__);
		$class = ref($msg) if $class eq __PACKAGE__;
		unshift @trace, @{$msg->trace};
		$msg = $msg->msg;
	}
	my $self = $class->new($msg);
	$tracelevel = 0 unless defined($tracelevel);
	$tracelevel++;
	my @traceback = traceback($tracelevel);
	unless (@trace)
	{
		unshift @trace, @traceback;
	} else
	{
		unshift @trace, $traceback[0];
	}
	$self->trace(\@trace);
	die $self;
}
################################################################################

sub new
{
	my $class = shift;
	my ($msg) = @_;
	my $self = $class->SUPER();
	$self->msg($msg);
	$self->debug((defined($main::DEBUG) and $main::DEBUG)? 1: 0);
	$self->trace([]);
	return $self;
}

sub message
{
	local $/ = "\n" unless defined($/);
	my $self = shift;
	my ($debug) = @_;
	$debug = $self->debug unless defined($debug);
	my $msg = $self->msg;
	my $result = "";
	$result .= "$msg$/" if defined($msg) and not ref($msg);
	return $result unless $debug;
	$result .= dump_trace(@{$self->trace});
	return $result;
}


1;
__END__
=head1 REPOSITORY

B<GitHub> L<https://github.com/orkunkaraduman/p5-Object-Base>

B<CPAN> L<https://metacpan.org/release/Object-Base>

=head1 SEE ALSO

=over

=item *

L<Object::Base|https://metacpan.org/pod/Object::Base>

=back

=head1 AUTHOR

Orkun Karaduman <orkunkaraduman@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2017  Orkun Karaduman <orkunkaraduman@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut


lib/Object/Base.pm
package Object::Base;
=head1 NAME

Object::Base - Multi-threaded base class to establish a class deriving relationship with parent classes

=head1 VERSION

version 1.13

=head1 ABSTRACT

Multi-threaded base class to establish a class deriving relationship with parent classes

	package Foo;
	use Object::Base;
	attributes ':shared', 'attr1', 'attr2';
	#
	package Bar;
	use Object::Base 'Foo';
	my $attr3_def = 6;
	my $attr3_val;
	attributes ':shared' => undef, 'attr2' => undef, ':lazy',
		'attr3' => {
			'default' => sub {
				my ($self, $key) = @_;
				print "> default key=$key\n";
				$attr3_val = $attr3_def-1;
				return $attr3_def;
			},
			'getter' => sub {
				my ($self, $key) = @_;
				print "> getter key=$key\n";
				return $attr3_val+1;
			},
			'setter' => sub {
				my ($self, $key, $value) = @_;
				print "> setter key=$key value=$value\n";
				$attr3_val = $value-1;
			},
		}
	;
	#
	package main;
	use threads;
	use threads::shared;
	#
	# object of Foo
	my $foo = Foo->new();
	#
	# special attribute ':shared'
	print "\$foo is ", is_shared($foo)? "shared": "not shared", "\n";
	#
	# usage of attribute
	$foo->attr1(1);
	print $foo->attr1, "\n"; # prints '1'
	#
	# attributes are lvalued
	$foo->attr1++;
	print $foo->attr1, "\n"; # prints '2'
	#
	# assigning ref values to shared class attributes
	eval { $foo->attr2 = { key1 => 'val1' } }; print "Eval: $@"; # prints error 'Eval: Invalid value for shared scalar at ...'
	$foo->attr2({ key2 => 'val2' }); # uses shared_clone assigning ref value
	print $foo->attr2->{key2}, "\n"; # prints 'val2'
	#
	# object of derived class Bar
	my $bar = Bar->new();
	#
	# features are overridable
	print "\$bar is ", is_shared($bar)? "shared": "not shared", "\n"; # prints '$bar is not shared'
	#
	# attributes can be added derived classes
	# attributes can have modifiers: default
	print "attr3 default value is ", $bar->attr3, "\n"; # prints 'attr3 default value is 6'
	#
	# attributes can have modifiers: setter
	$bar->attr3 = 3;
	print "attr3 set", "\n";
	#
	# attributes can have modifiers: getter
	print "attr3 value ", $bar->attr3, " and stored as $attr3_val", "\n"; # prints 'attr3 value 3 and stored as 2'
	#
	# attributes are inheritable
	$bar->attr1(3);
	#
	# attributes are overridable
	eval { $bar->attr2 = 4 }; print "Eval: $@"; # prints error 'Eval: Attribute attr2 is not defined in Bar at ...'
	#
	# attributes in thread
	my $thr1 = threads->create(sub { $foo->attr1 = 5; $bar->attr1 = 5; });
	my $thr2 = threads->create(sub { sleep 1; print "\$foo is shared and attr1: ", $foo->attr1, ", \$bar is not shared and attr1: ", $bar->attr1, "\n"; });
	# prints '$foo is shared and attr1: 5, $bar is not shared and attr1: 3'
	$thr1->join();
	$thr2->join();

=head1 DESCRIPTION

Object::Base provides blessed and thread-shared(with :shared feature) object with in B<new> method. B<new> method
can be used as a constructor and overridable in derived classes. B<new()> should be called in derived class
constructors to create and bless self-object.

Derived classes own package automatically uses threads, threads::shared, strict, warnings with using Object::Base. If
Perl is not built to support threads; it uses forks, forks::shared instead of threads, threads::shared. Object::Base
should be loaded as first module.

Import parameters of Object::Base, define parent classes of derived class.
If none of parent classes derived from Object::Base or any parent isn't defined, Object::Base is automatically added
in parent classes.

=head2 Attributes

Attributes define read-write accessors binded value of same named key in objects own hash if attribute names is
valid subroutine identifiers. Otherwise, attribute defines B<feature> to get new features into class.

Attributes;

=over

=item *

Lvaluable

=item *

Inheritable

=item *

Overridable

=item *

Redefinable

=back

=head3 Modifiers

Attributes can have their own modifiers in hash reference at definition. Attribute modifiers don't work with ':shared' feature.

=head4 default

getter method of default value of attribute, otherwise value is default value

	attributes
		'attr1' => {
			'default' => sub {
				my ($self, $key) = @_;
				return "default value of $key";
			},
		},
		'attr2' => {
			'default' => "default value of attr2",
		};

=head4 getter

getter method of attribute

	my $attr1_val;
	attributes
		'attr1' => {
			'getter' => sub {
				my ($self, $key) = @_;
				return $attr1_val;
			},
		};

=head4 setter

setter method of attribute

	my $attr1_val;
	attributes
		'attr1' => {
			'setter' => sub {
				my ($self, $key, $value) = @_;
				$attr1_val = $value;
			},
		};

=head3 Features

=head4 :shared

Class will be craated as thread-shared. But attribute modifiers don't work with ':shared' feature.

=head4 :lazy

Attributes will be initialized with default values using default modifier at first fetching or storing instead of object construction with new().

=cut
BEGIN
{
	require Config;
	if ($Config::Config{'useithreads'})
	{
		require threads;
		threads->import();
		require threads::shared;
		threads::shared->import();
	} else
	{
		require forks;
		forks->import();
		require forks::shared;
		forks::shared->import();
	}
}
use strict;
no strict qw(refs);
use warnings;


BEGIN
{
	require 5.008;
	$Object::Base::VERSION = '1.13';
	$Object::Base::ISA = ();
}


my $package = __PACKAGE__;
my $context = $package;
$context =~ s/\Q::\E//g;


sub import
{
	die "$package can not be imported at run-time" if ${^GLOBAL_PHASE} eq "RUN";
	my $importer = shift;
	my $caller = caller;
	return unless $importer eq $package;
	eval join "\n",
		"package $caller;",
		<< "EOF",
BEGIN
{
	require Config;
	if (\$Config::Config{'useithreads'})
	{
		require threads;
		threads->import();
		require threads::shared;
		threads::shared->import();
	} else
	{
		require forks;
		forks->import();
		require forks::shared;
		forks::shared->import();
	}
}
use SUPER;
EOF
		"use strict;",
		"use warnings;",
		"",
		(exists(&{"${caller}::attributes"})? "": "sub attributes { ${package}::attributes(\@_) }"),
		"",
		"\%${caller}::${context} = () unless defined(\\\%${caller}::${context});",
		"",
		(
			map {
				my $p = (defined and not ref)? $_: "";
				$p and /^[^\W\d]\w*(\:\:[^\W\d]\w*)*\z/s or die "Invalid base-class name $p";
				<< "EOF";
eval { require $_ };
push \@${caller}::ISA, '$_';
if ($_->isa('$package'))
{
	\$${caller}::${context}{\$_} = \$$_\::${context}{\$_} for (keys \%$_\::${context});
}
EOF
			} @_
		),
		"push \@${caller}::ISA, '$package' unless UNIVERSAL::isa('${caller}', '$package');";
	die "Failed to import $package in $caller: $@" if $@;
	return 1;
}

sub attributes
{
	my $caller = caller;
	die "$caller is not $package class" unless UNIVERSAL::isa($caller, $package);
	%{"${caller}::${context}"} = () unless defined(\%{"${caller}::${context}"});
	my $l;
	for (@_)
	{
		if (not defined($_) or ref($_))
		{
			next if not defined($l) or ref($l);
			${"${caller}::${context}"}{$l} = $_;
			next;
		}
		${"${caller}::${context}"}{$_} = {};
	} continue
	{
		$l = $_;
	}
	eval join "\n",
		"package $caller;",
		"",
		map {
			<< "EOF";
sub $_ :lvalue
{
	my \$self = shift;
	die 'Attribute $_ is not defined in $caller' unless \$${caller}::${context}{"$_"};
	die 'Object is not derived from $package' unless defined(\$self) and UNIVERSAL::isa(ref(\$self), '$package');
	my \@args = \@_;
	if (\@args >= 1)
	{
		unless (is_shared(%{\$self}) and ref(\$args[0]) and not is_shared(\$args[0]))
		{
			\$self->{"$_"} = \$args[0];
		} else
		{
			\$self->{"$_"} = shared_clone(\$args[0]);
		}
	}
	\$self->{"$_"};
}
EOF
		} grep { /^[^\W\d]\w*\z/s and not exists(&{"${caller}::$_"}) } keys %{"${caller}::${context}"};
	die "Failed to generate attributes in $caller: $@" if $@;
	return 1;
}

sub new
{
	my $class = shift;
	die "Invalid $package class" unless
		defined($class) and
		not ref($class) and
		UNIVERSAL::isa($class, $package);
	die "$package context is not defined" unless defined(\%{"${class}::${context}"});
	my $self = {};
	tie %$self, "${package}::TieHash", $class, \$self unless ${"${class}::${context}"}{":shared"};
	$self = shared_clone($self) if ${"${class}::${context}"}{":shared"};
	bless $self, $class;
}

sub DESTROY
{
}


package Object::Base::TieHash;
BEGIN
{
	require Config;
	if ($Config::Config{'useithreads'})
	{
		require threads;
		threads->import();
		require threads::shared;
		threads::shared->import();
	} else
	{
		require forks;
		forks->import();
		require forks::shared;
		forks::shared->import();
	}
}
use strict;
no strict qw(refs);
use warnings;


BEGIN
{
	require 5.008;
	$Object::Base::TieHash::VERSION = $Object::Base::VERSION;
}


sub TIEHASH
{
	my $class = shift;
	my $self = [{}, @_, {}];
	bless $self, $class;
	for (grep /^[^\W\d]\w*\z/s, keys(%{"$self->[1]\::${context}"}))
	{
		$self->[0]->{$_} = undef;
		$self->def($_) unless ${"$self->[1]\::${context}"}{":lazy"};
	}
	$self;
}

sub STORE
{
	my $self = shift;
	my ($key, $value) = @_;
	if ($key =~ /^[^\W\d]\w*\z/s)
	{
		$self->def($key);
		my $attr = ${"$self->[1]\::${context}"}{$key};
		if (ref($attr) eq 'HASH' and exists($attr->{"setter"}))
		{
			my $setter = $attr->{"setter"};
			if (ref($setter) eq 'CODE')
			{
				$setter->(${$self->[2]}, $key, $value);
			}
		}
	}
	$self->[0]->{$key} = $value;
}

sub FETCH
{
	my $self = shift;
	my ($key) = @_;
	if ($key =~ /^[^\W\d]\w*\z/s)
	{
		$self->def($key);
		my $attr = ${"$self->[1]\::${context}"}{$key};
		if (ref($attr) eq 'HASH' and exists($attr->{"getter"}))
		{
			my $getter = $attr->{"getter"};
			if (ref($getter) eq 'CODE')
			{
				$self->[0]->{$key} = $getter->(${$self->[2]}, $key);
			}
		}
	}
	$self->[0]->{$key};
}

sub EXISTS
{
	exists $_[0][0]->{$_[1]};
}

sub DELETE
{
	delete $_[0][$#{$_[0]}]->{$_[1]};
	delete $_[0][0]->{$_[1]};
}

sub CLEAR
{
	%{$_[0][$#{$_[0]}]} = ();
	%{$_[0][0]} = ();
}

sub FIRSTKEY
{
	my $a = scalar keys %{$_[0][0]};
	each %{$_[0][0]};
}

sub NEXTKEY
{
	each %{$_[0][0]};
}

sub SCALAR
{
	scalar %{$_[0][0]};
}

sub def
{
	my $self = shift;
	my ($key) = @_;
	unless (exists($self->[$#{$self}]->{$key}))
	{
		my $val = undef;
		my $attr = ${"$self->[1]\::${context}"}{$key};
		if (ref($attr) eq 'HASH' and exists($attr->{"default"}))
		{
			my $default = $attr->{"default"};
			if (ref($default) eq 'CODE')
			{
				$val = $default->(${$self->[2]}, $key);
			} else
			{
				$val = $default;
			}
		}
		$self->[$#{$self}]->{$key} = $val;
		$self->[0]->{$key} = $val;
	}
	return $self->[$#{$self}]->{$key};
}


1;
__END__
=head1 REPOSITORY

B<GitHub> L<https://github.com/orkunkaraduman/p5-Object-Base>

B<CPAN> L<https://metacpan.org/release/Object-Base>

=head1 SEE ALSO

=over

=item *

L<Object::Exception|https://metacpan.org/pod/Object::Exception>

=back

=head1 AUTHOR

Orkun Karaduman <orkunkaraduman@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2017  Orkun Karaduman <orkunkaraduman@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

=cut
